File: javascript/kyne.js
/**
 * Kyne JavaScript SDK
 * Official JavaScript/Node.js SDK for Kyne Payment Verification Gateway
 */

class KyneError extends Error {
    constructor(message, statusCode = null) {
        super(message);
        this.name = 'KyneError';
        this.statusCode = statusCode;
    }
}

class AuthenticationError extends KyneError {
    constructor(message) {
        super(message, 401);
        this.name = 'AuthenticationError';
    }
}

class ValidationError extends KyneError {
    constructor(message) {
        super(message, 400);
        this.name = 'ValidationError';
    }
}

class Kyne {
    /**
     * Create a Kyne client
     * @param {string} apiKey - Your secret API key (sk_test_xxx or sk_live_xxx)
     * @param {Object} options - Optional configuration
     * @param {string} options.baseUrl - Custom API base URL
     * @param {number} options.timeout - Request timeout in milliseconds
     */
    constructor(apiKey, options = {}) {
        if (!apiKey) {
            throw new AuthenticationError('API key is required');
        }
        
        if (!apiKey.startsWith('sk_test_') && !apiKey.startsWith('sk_live_')) {
            throw new AuthenticationError('Invalid API key format. Must start with sk_test_ or sk_live_');
        }
        
        this.apiKey = apiKey;
        this.baseUrl = (options.baseUrl || 'https://api.kyne.com').replace(/\/$/, '');
        this.timeout = options.timeout || 30000;
        this.mode = apiKey.startsWith('sk_test_') ? 'test' : 'live';
    }
    
    /**
     * Verify a payment transaction
     * @param {Object} params - Verification parameters
     * @param {string} params.transactionId - Bank transaction reference
     * @param {number} params.amount - Expected amount in ETB
     * @param {string} [params.provider] - Payment provider (cbe, telebirr). Auto-detected if not provided.
     * @param {string} [params.merchantName] - Your registered bank account name
     * @returns {Promise<Object>} Verification result
     * 
     * @example
     * const result = await client.verify({
     *     transactionId: 'FT24352648751234',
     *     amount: 100,
     *     provider: 'cbe'
     * });
     */
    async verify(params) {
        const { transactionId, amount, provider, merchantName, subProvider } = params;
        
        if (!transactionId) {
            throw new ValidationError('transactionId is required');
        }
        if (!amount) {
            throw new ValidationError('amount is required');
        }
        
        // Auto-detect provider
        let detectedProvider = provider;
        if (!detectedProvider) {
            detectedProvider = transactionId.toUpperCase().startsWith('FT') ? 'cbe' : 'telebirr';
        }
        
        const data = new URLSearchParams();
        data.append('provider', detectedProvider);
        data.append('transaction_id', transactionId);
        data.append('amount', amount.toString());
        data.append('merchant_name', merchantName || 'Kyne Verification');
        
        if (subProvider) {
            data.append('sub_provider', subProvider);
        }
        
        return this._request('POST', '/api/v1/verify', data);
    }
    
    /**
     * Quick verification with auto-detected provider
     * @param {string} transactionId - Bank transaction reference
     * @param {number} amount - Expected amount
     * @returns {Promise<Object>} Verification result
     */
    async quickVerify(transactionId, amount) {
        const data = new URLSearchParams();
        data.append('transaction_id', transactionId);
        data.append('amount', amount.toString());
        
        return this._request('POST', '/api/v1/quick-verify', data);
    }
    
    /**
     * Get transaction history
     * @param {number} [limit=50] - Maximum number of transactions
     * @returns {Promise<Array>} List of transactions
     */
    async getHistory(limit = 50) {
        return this._request('GET', '/api/v1/history');
    }
    
    /**
     * Create a webhook endpoint
     * @param {Object} params - Webhook parameters
     * @param {string} params.url - Webhook URL
     * @param {Array<string>} params.events - Events to subscribe to
     * @returns {Promise<Object>} Created webhook with secret
     */
    async createWebhook(params) {
        const { url, events } = params;
        
        if (!url) {
            throw new ValidationError('url is required');
        }
        if (!events || events.length === 0) {
            throw new ValidationError('events are required');
        }
        
        return this._request('POST', '/api/v1/webhooks/', {
            url,
            events
        }, true);
    }
    
    /**
     * List all webhooks
     * @returns {Promise<Array>} List of webhooks
     */
    async listWebhooks() {
        return this._request('GET', '/api/v1/webhooks/');
    }
    
    /**
     * Delete a webhook
     * @param {string} webhookId - Webhook ID to delete
     * @returns {Promise<Object>} Deletion result
     */
    async deleteWebhook(webhookId) {
        return this._request('DELETE', `/api/v1/webhooks/${webhookId}`);
    }
    
    /**
     * Internal request method
     */
    async _request(method, path, data = null, isJson = false) {
        const url = `${this.baseUrl}${path}`;
        
        const headers = {
            'Authorization': `Bearer ${this.apiKey}`,
            'User-Agent': 'Kyne-JavaScript-SDK/1.0'
        };
        
        const options = {
            method,
            headers
        };
        
        if (data) {
            if (isJson) {
                headers['Content-Type'] = 'application/json';
                options.body = JSON.stringify(data);
            } else {
                headers['Content-Type'] = 'application/x-www-form-urlencoded';
                options.body = data.toString();
            }
        }
        
        try {
            const controller = new AbortController();
            const timeoutId = setTimeout(() => controller.abort(), this.timeout);
            
            const response = await fetch(url, {
                ...options,
                signal: controller.signal
            });
            
            clearTimeout(timeoutId);
            
            if (response.status === 401) {
                throw new AuthenticationError('Invalid API key');
            }
            
            if (response.status === 400) {
                const error = await response.json();
                throw new ValidationError(error.detail || 'Validation error');
            }
            
            if (response.status >= 500) {
                throw new KyneError('Server error', response.status);
            }
            
            return response.json();
            
        } catch (error) {
            if (error.name === 'AbortError') {
                throw new KyneError('Request timed out');
            }
            if (error instanceof KyneError) {
                throw error;
            }
            throw new KyneError(`Request failed: ${error.message}`);
        }
    }
}

/**
 * Verify webhook signature
 * @param {string} payload - Raw request body
 * @param {string} signature - X-Kyne-Signature header value
 * @param {string} secret - Your webhook secret
 * @returns {boolean} Whether signature is valid
 */
async function verifyWebhookSignature(payload, signature, secret) {
    // For Node.js
    if (typeof require !== 'undefined') {
        const crypto = require('crypto');
        const expected = crypto
            .createHmac('sha256', secret)
            .update(payload)
            .digest('hex');
        return `sha256=${expected}` === signature;
    }
    
    // For browser with Web Crypto API
    const encoder = new TextEncoder();
    const keyData = encoder.encode(secret);
    const data = encoder.encode(payload);
    
    const key = await crypto.subtle.importKey(
        'raw',
        keyData,
        { name: 'HMAC', hash: 'SHA-256' },
        false,
        ['sign']
    );
    
    const signatureBuffer = await crypto.subtle.sign('HMAC', key, data);
    const expected = Array.from(new Uint8Array(signatureBuffer))
        .map(b => b.toString(16).padStart(2, '0'))
        .join('');
    
    return `sha256=${expected}` === signature;
}

// Export for different module systems
if (typeof module !== 'undefined' && module.exports) {
    // CommonJS
    module.exports = { Kyne, KyneError, AuthenticationError, ValidationError, verifyWebhookSignature };
} else if (typeof window !== 'undefined') {
    // Browser
    window.Kyne = Kyne;
    window.verifyWebhookSignature = verifyWebhookSignature;
}

// ES Modules
export { Kyne, KyneError, AuthenticationError, ValidationError, verifyWebhookSignature };
export default Kyne;

File: python/kyne.py
"""
Kyne Python SDK
Official Python SDK for Kyne Payment Verification Gateway
"""

import requests
from typing import Optional, Dict, Any, List
from dataclasses import dataclass
from enum import Enum


class Provider(Enum):
    """Supported payment providers"""
    CBE = "cbe"
    TELEBIRR = "telebirr"
    BANK_TRANSFER = "bank_transfer"


@dataclass
class VerificationResult:
    """Result of a payment verification"""
    valid: bool
    status: str
    provider: Optional[str] = None
    transaction_id: Optional[str] = None
    amount: Optional[float] = None
    reason: Optional[str] = None
    mode: Optional[str] = None
    payer: Optional[str] = None
    
    @classmethod
    def from_response(cls, data: Dict[str, Any]) -> 'VerificationResult':
        return cls(
            valid=data.get('valid', False),
            status=data.get('status', 'unknown'),
            provider=data.get('provider'),
            transaction_id=data.get('transaction_id'),
            amount=data.get('amount'),
            reason=data.get('reason'),
            mode=data.get('mode'),
            payer=data.get('payer')
        )


@dataclass 
class Transaction:
    """A verified transaction record"""
    id: str
    provider: str
    external_id: str
    amount: float
    status: str
    created_at: str
    mode: str


class KyneError(Exception):
    """Base exception for Kyne SDK"""
    def __init__(self, message: str, status_code: int = None):
        self.message = message
        self.status_code = status_code
        super().__init__(self.message)


class AuthenticationError(KyneError):
    """Raised when API key is invalid"""
    pass


class ValidationError(KyneError):
    """Raised when request validation fails"""
    pass


class Kyne:
    """
    Kyne Payment Verification Client
    
    Usage:
        client = Kyne(api_key="sk_test_xxx")
        result = client.verify(provider="cbe", transaction_id="FT123", amount=100)
    """
    
    DEFAULT_BASE_URL = "https://api.kyne.com"
    
    def __init__(
        self, 
        api_key: str,
        base_url: Optional[str] = None,
        timeout: int = 30
    ):
        """
        Initialize Kyne client.
        
        Args:
            api_key: Your secret API key (sk_test_xxx or sk_live_xxx)
            base_url: Optional custom API base URL
            timeout: Request timeout in seconds
        """
        if not api_key:
            raise AuthenticationError("API key is required")
        
        if not api_key.startswith(("sk_test_", "sk_live_")):
            raise AuthenticationError("Invalid API key format. Must start with sk_test_ or sk_live_")
        
        self.api_key = api_key
        self.base_url = (base_url or self.DEFAULT_BASE_URL).rstrip('/')
        self.timeout = timeout
        self.mode = "test" if api_key.startswith("sk_test_") else "live"
        
        self._session = requests.Session()
        self._session.headers.update({
            'Authorization': f'Bearer {api_key}',
            'Content-Type': 'application/x-www-form-urlencoded',
            'User-Agent': 'Kyne-Python-SDK/1.0'
        })
    
    def verify(
        self,
        transaction_id: str,
        amount: float,
        provider: str = None,
        merchant_name: str = None,
        sub_provider: str = None
    ) -> VerificationResult:
        """
        Verify a payment transaction.
        
        Args:
            transaction_id: Bank transaction reference (e.g., FT123456789)
            amount: Expected amount in ETB
            provider: Payment provider (cbe, telebirr, bank_transfer). Auto-detected if not provided.
            merchant_name: Your registered bank account name
            sub_provider: Sub-provider for bank transfers (e.g., Payoneer)
            
        Returns:
            VerificationResult with valid=True if payment is verified
            
        Example:
            result = client.verify(
                provider="cbe",
                transaction_id="FT24352648751234",
                amount=100,
                merchant_name="My Shop"
            )
        """
        # Auto-detect provider if not specified
        if not provider:
            if transaction_id.upper().startswith("FT"):
                provider = "cbe"
            else:
                provider = "telebirr"
        
        data = {
            'provider': provider,
            'transaction_id': transaction_id,
            'amount': amount,
        }
        
        if merchant_name:
            data['merchant_name'] = merchant_name
        else:
            data['merchant_name'] = 'Kyne Verification'
            
        if sub_provider:
            data['sub_provider'] = sub_provider
        
        response = self._request('POST', '/api/v1/verify', data=data)
        return VerificationResult.from_response(response)
    
    def quick_verify(self, transaction_id: str, amount: float) -> VerificationResult:
        """
        Quick verification with auto-detected provider.
        
        Automatically detects CBE (FT...) or Telebirr from transaction ID format.
        
        Args:
            transaction_id: Bank transaction reference
            amount: Expected amount
            
        Returns:
            VerificationResult
        """
        data = {
            'transaction_id': transaction_id,
            'amount': amount
        }
        
        response = self._request('POST', '/api/v1/quick-verify', data=data)
        return VerificationResult.from_response(response)
    
    def get_history(self, limit: int = 50) -> List[Transaction]:
        """
        Get transaction history.
        
        Args:
            limit: Maximum number of transactions to return
            
        Returns:
            List of Transaction objects
        """
        response = self._request('GET', '/api/v1/history')
        
        transactions = []
        for tx in response:
            transactions.append(Transaction(
                id=tx.get('id'),
                provider=tx.get('provider'),
                external_id=tx.get('external_id'),
                amount=tx.get('amount'),
                status=tx.get('status'),
                created_at=tx.get('created_at'),
                mode=tx.get('mode')
            ))
        
        return transactions
    
    def _request(
        self, 
        method: str, 
        path: str, 
        data: Dict = None,
        params: Dict = None
    ) -> Dict[str, Any]:
        """Make API request"""
        url = f"{self.base_url}{path}"
        
        try:
            response = self._session.request(
                method=method,
                url=url,
                data=data,
                params=params,
                timeout=self.timeout
            )
            
            if response.status_code == 401:
                raise AuthenticationError("Invalid API key", status_code=401)
            
            if response.status_code == 400:
                error = response.json()
                raise ValidationError(error.get('detail', 'Validation error'), status_code=400)
            
            if response.status_code >= 500:
                raise KyneError("Server error", status_code=response.status_code)
            
            return response.json()
            
        except requests.exceptions.Timeout:
            raise KyneError("Request timed out")
        except requests.exceptions.ConnectionError:
            raise KyneError("Connection error")


# Convenience function
def verify(api_key: str, transaction_id: str, amount: float, **kwargs) -> VerificationResult:
    """
    Quick verification without creating a client.
    
    Usage:
        import kyne
        result = kyne.verify("sk_test_xxx", "FT123456", 100)
    """
    client = Kyne(api_key)
    return client.verify(transaction_id=transaction_id, amount=amount, **kwargs)


__version__ = "1.0.0"
__all__ = ['Kyne', 'VerificationResult', 'Transaction', 'Provider', 'KyneError', 'verify']
